ibuffer-project.el

```diff
@@ -63,18 +63,34 @@
 (require 'ibuffer)
 (require 'ibuf-ext)
 
+(defvar ibuffer-project-group-cache (make-hash-table)
+  "Cache for mappings from a directory to its group data.")
+
+(defun ibuffer-project-purge-group-cache ()
+  "Purge the group cache."
+  (setq ibuffer-project-group-cache (make-hash-table)))
+
 (defun ibuffer-project-root (buf)
   "Return a cons cell (project-root . root-type) for BUF."
   (unless (string-match-p "^ " (buffer-name buf))
-    (with-current-buffer buf
-      (let ((dir (cdr (project-current))))
-        (cond
-         (dir (cons dir 'project))
-         (default-directory (cons (abbreviate-file-name default-directory) 'directory)))))))
+    (let ((dir (buffer-local-value 'default-directory buf)))
+      (when dir
+        (or (gethash dir ibuffer-project-group-cache)
+            (let* ((root (cdr (project-current nil dir)))
+                   (result (cond
+                            (root (cons root 'project))
+                            (dir (cons (abbreviate-file-name dir) 'directory)))))
+              (when result
+                (puthash dir result ibuffer-project-group-cache))
+              result))))))
 
 (defun ibuffer-project-group-name (root type)
   "Return group name for project ROOT and TYPE."
-  (format "%s: %s" (if (eq type 'project) "Project" "Directory") root))
+  (format "%s: %s"
+          (cl-case type
```

Comment:
How `cl-case` is better than `if` here? If this change is not required I'd rather leave it as is.

Reply:
You are right, but I am going to use it in tramp branch.